src/suricata-plugin.h

```diff
@@ -45,12 +45,22 @@ typedef struct SCPlugin_ {
  * Structure used to define a file type plugin.
  *
  * Currently only used by the Eve output type.
+ *
+ * name -- The plugin name. This name is used to identify the plugin: eve-log.filetype and in the
+ * plugins: section
  */
 typedef struct SCPluginFileType_ {
     char *name;
-    int (*Open)(ConfNode *conf, void **data);
-    int (*Write)(const char *buffer, int buffer_len, void *ctx);
-    void (*Close)(void *ctx);
+    /* Init Called on first access */
+    int (*Init)(ConfNode *conf, bool threaded, void **init_data);
+    /* Write - Called on each write to the object */
+    int (*Write)(const char *buffer, int buffer_len, void *init_data, void *thread_data);
+    /* Close - Called on final close */
+    void (*Deinit)(void *init_data);
+    /* ThreadInit - Called for each thread using file object*/
+    int (*ThreadInit)(void *init_data, int thread_id, void **thread_data);
```

Comment:
Seeing Victor's comments on thread local storage, is this needed? Or could thread_init just stuff its data in thread local storage?

Reply:
We could use TLS in lieu and yes, for the second question. Both come with extra work, which is fine.

Strictly speaking, the changes victor proposed are somewhat orthogonal but related as there as lots of places in suri that _could be_ modified to use tls over the pattern used here (and elsewhere in suri).